### ISSUE TITLE
`is_full` without decoding whole collection

### ISSUE THREAD

**Issue by bragov4ik** [paritytech/substrate#14725]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

In storages with bounded collections as elements sometimes it is necessary to check if the collection is full or not. For example, it seems better to check such things before performing some expensive computations and avoid unnecessary work by returning an error right away.

For example instead of failing late like here:
```rust
// <heavy computation>
// ***
// </heavy computation>
<Storage<T>>::mutate(
    key1,
    |list| list.try_push()
).map_err(|_| Error::StorageIsFull)?;
```
one could write something like this
```rust
if <Storage<T>>::decode_is_full(key1) {
    return Err(Error::StorageIsFull);
}
// <heavy computation>
// ***
// </heavy computation>
<Storage<T>>::mutate(
    key1,
    |list| list.try_push()
).map_err(|_| Error::StorageIsFull)?;
```
It is even more useful when such prechecks/preconditions are combined into a single function that's called at the start

### Request

Add `decode_is_full` functions to storages that check if the bound of the value is reached.

### Solution

As I see, it can be accomplished using [`StorageDecodeLength`](https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageDecodeLength.html) trait. For example, new trait `StorageDecodeIsFull` can be created:

```rust
pub trait StorageDecodeIsFull: private::Sealed + StorageDecodeLength {
	fn decode_is_full(key: &[u8]) -> Option<bool> {...}
}
```

It seems that it's possible to use `StorageDecodeLength::decode_len()` in the method and compare it with bound of the value's type.

In other words, it can be done just like [`is_full` from `BoundedVec`](https://github.com/paritytech/parity-common/blob/d3a9327124a66e52ca1114bb8640c02c18c134b8/bounded-collections/src/bounded_vec.rs#L440):
```rust
let bound = ...;
decode_len(key) >= bound
```

The only unclear part that I see is how to get the bound, but it should be doable, since it's a compile-time constant.

### Are you willing to help with this request?

Yes!
